File: mini/tools/pathlint.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
pathlint —— 走位路径体检

用户问「要怎么才能彻底避免」角色从家具上走过去。答案是：不能靠眼睛盯，要靠这个。

角色穿模有三个来源，只修第三个是不够的：
  ① 锚点落在家具里        → 终点就在柜子里（roomstats 之外，本工具也查）
  ② 路网【边】穿过家具     → 路本身画在柜子里，走多少次穿多少次 ★最隐蔽
  ③ 最后一段直奔锚点      → 从节点到锚点那条直线不经过任何避障

阿云房用 gridPath（网格 BFS，自动绕障），桃桃房用 NAV 路网（手工节点 + 直线边）。
路网的边是人手连的，没有任何机制保证它不穿家具 —— 这正是 ② 的温床。

用法：python3 tools/pathlint.py <design.html>
"""
import re, sys

def seg_rect(x1, y1, x2, y2, rx, ry, rw, rh):
    """线段与矩形是否相交（含包含）"""
    if max(x1, x2) < rx or min(x1, x2) > rx + rw:
        if not (rx <= x1 <= rx+rw or rx <= x2 <= rx+rw): return False
    def side(px, py): return (px - rx, py - ry)
    # 粗筛：包围盒不交则不交
    if max(x1,x2) < rx or min(x1,x2) > rx+rw or max(y1,y2) < ry or min(y1,y2) > ry+rh:
        return False
    # 端点在矩形内
    for px, py in ((x1,y1),(x2,y2)):
        if rx <= px <= rx+rw and ry <= py <= ry+rh: return True
    # 线段与四条边求交
    def inter(ax,ay,bx,by,cx,cy,dx,dy):
        d = (bx-ax)*(dy-cy) - (by-ay)*(dx-cx)
        if d == 0: return False
        t = ((cx-ax)*(dy-cy) - (cy-ay)*(dx-cx)) / d
        u = ((cx-ax)*(by-ay) - (cy-ay)*(bx-ax)) / d
        return 0 <= t <= 1 and 0 <= u <= 1
    E = [(rx,ry,rx+rw,ry), (rx+rw,ry,rx+rw,ry+rh), (rx+rw,ry+rh,rx,ry+rh), (rx,ry+rh,rx,ry)]
    return any(inter(x1,y1,x2,y2,*e) for e in E)

def load(path):
    s = open(path, encoding='utf-8').read()
    A = {}
    for m in re.finditer(r'def\(["\']([a-z0-9_]+)["\'],\s*\{(.{0,320})', s, re.S):
        b = m.group(2)
        w = re.search(r'\bw:\s*(\d+)', b); h = re.search(r'\bh:\s*(\d+)', b)
        f = re.search(r'foot:\s*\[\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)', b)
        walk = 'walkable' in b            # 坐垫/门垫能踩，不算障碍
        A[m.group(1)] = (int(w.group(1)) if w else 0, int(h.group(1)) if h else 0,
                         (0,0,0,0) if walk else (tuple(int(f.group(i)) for i in (1,2,3,4)) if f else (0,0,0,0)))
    return s, A

def furniture(s, A, key):
    m = re.search(r'window\.' + key + r'\s*=\s*\{', s)
    if not m: return []
    j = s.find('plan:', m.end()); k = s.find('\n  }', j)
    out = []
    for mm in re.finditer(r"\['([a-z0-9_]+)',\s*(\d+),\s*(\d+)", s[j:k]):
        n, x, y = mm.group(1), int(mm.group(2)), int(mm.group(3))
        w, h, f = A.get(n, (0,0,(0,0,0,0)))
        if f == (0,0,0,0): continue          # 挂件/地毯不占地
        out.append((n, x + f[0], y + f[1], f[2], f[3]))
    return out

FOOT = (48, 136)   # 锚点 → 脚点偏移

def check(path, key, canvas, label):
    s, A = load(path)
    furn = furniture(s, A, key)
    i = s.find("getElementById('%s')" % canvas)
    seg = s[i:s.find('</script>', i)] if i >= 0 else ''
    bad = 0

    nav = re.search(r'const NAV\s*=\s*\{([^}]*)\}', seg)
    edges = re.search(r'const EDGES\s*=\s*\{([^}]*(?:\}[^}]*)*?)\}\s*\n', seg)
    print('\n═══ %s ═══' % label)
    if not nav:
        print('  用网格寻路（gridPath），边由网格保证避障 —— 无路网可查')
    else:
        N = {k: (int(x), int(y)) for k, x, y in re.findall(r'(\w+):\s*\[(\d+),\s*(\d+)\]', nav.group(1))}
        print('  路网 %d 节点' % len(N))
        seen = set()
        for a, lst in re.findall(r'(\w+):\s*\[([^\]]*)\]', edges.group(1) if edges else ''):
            for b in re.findall(r"'(\w+)'", lst):
                if (b, a) in seen or a not in N or b not in N: continue
                seen.add((a, b))
                x1, y1 = N[a]; x2, y2 = N[b]
                for n, fx, fy, fw, fh in furn:
                    if seg_rect(x1, y1, x2, y2, fx, fy, fw, fh):
                        print('  ✗ 边 %s→%s 穿过 %s' % (a, b, n)); bad += 1
        # 节点 → 锚点 的最后一段
        for m in re.finditer(r"\{\s*id:\s*'(\w+)',[^{}]*node:\s*'(\w+)'[^{}]*x:\s*(\d+),\s*y:\s*(\d+)[^{}]*\}", seg):
            aid, nd, ax, ay = m.group(1), m.group(2), int(m.group(3)), int(m.group(4))
            if nd not in N: continue
            x1, y1 = N[nd]; x2, y2 = ax + FOOT[0], ay + FOOT[1]
            for n, fx, fy, fw, fh in furn:
                if seg_rect(x1, y1, x2, y2, fx, fy, fw, fh):
                    print('  ✗ 末段 %s(节点 %s → 锚点) 穿过 %s' % (aid, nd, n)); bad += 1
    # 锚点本身
    for m in re.finditer(r"\{\s*id:\s*'(\w+)',[^{}]*x:\s*(\d+),\s*y:\s*(\d+)[^{}]*\}", seg):
        if 'node:' not in m.group(0): continue
        if 'onTop' in m.group(0): continue      # 猫趴在桌上这类是设定，不是穿模
        fx2, fy2 = int(m.group(2)) + FOOT[0], int(m.group(3)) + FOOT[1]
        for n, fx, fy, fw, fh in furn:
            if fx <= fx2 <= fx + fw and fy <= fy2 <= fy + fh:
                print('  ✗ 锚点 %s 落在 %s 内' % (m.group(1), n)); bad += 1
    print('  %s' % ('✓ 无穿模路径' if bad == 0 else '共 %d 处' % bad))
    return bad

if __name__ == '__main__':
    p = sys.argv[1] if len(sys.argv) > 1 else 'mini/design.html'
    total = 0
    for key, cv, lab in [('AYUN_ROOM','ayunCanvas','阿云'), ('TAO_ROOM','taoCanvas','桃桃'),
                         ('POPO_ROOM','popoCanvas','婆婆'), ('TENZ_ROOM','tenzCanvas','丹增')]:
        total += check(p, key, cv, lab)
    print('\n合计 %d 处' % total)
    sys.exit(1 if total else 0)

File: mini/tools/pathlint.js
#!/usr/bin/env bun
/**
 * pathlint —— 行为锚点与路径体检(WORKFLOW S7 的门禁之一)
 *
 * WORKFLOW 从第一版就写着「pathlint 0 处」,文档里连实测数字都在
 * (桃桃 14 处穿模、阿云 4 处),但 tools/ 里【没有这个文件】——
 * 它随那次 137 commit 回退一起没了,门禁于是空写了很久:
 * 要求跑一支跑不了的工具,等于这一关谁都过。这是第六种假绿。
 *
 * 查四件事:
 *   ① 锚点站进家具里 ★ 最常见 —— 挪了家具没挪锚点,他就站到柜子里去了。
 *      `startWalkTo` 会【无条件】把锚点本身 push 进路径末尾,所以哪怕
 *      锚点是障碍格,他也会在最后一段直线走进去。寻路绕得开,这一步绕不开。
 *   ② 锚点够不着 —— 引擎实际把他放下的地方离锚点超过两格,他会停在半路。
 *   ③ 路径穿家具 —— gridPath 返回的折线有拐点落在障碍格上。
 *   ④ 表演锚点 —— perform.actor 的落点同样要查,那是玩家点按钮唯一会去的地方。
 *
 * 锚点语义有两种,不能混:主角的 act.x/y 是 **sprite 左上角**(要加 foot 偏移
 * 才是脚点),宠物的 act.x/y **本身就是脚点**。房间在 ROOM.acts 上用
 * anchorIsFoot 如实标注,与 startWalkTo 的 { foot:[0,0] } 一一对应。
 *
 * ⚠ 这一支报 0 **不等于**走位对了 —— 它建立在 S5 的 `foot` 之上。
 *   碰撞体画错了,烧出来的格就是错的,穿模照样过。先 assetlint + roomaudit D。
 *
 * 用法:bun tools/pathlint.js <design.html 绝对路径> [房间名]
 */
const { chromium } = require('playwright')
const FILE = process.argv[2], ONLY = process.argv[3]
if (!FILE) { console.error('用法: bun tools/pathlint.js <design.html 绝对路径> [房间名]'); process.exit(2) }

;(async () => {
  const b = await chromium.launch(), p = await b.newPage()
  await p.goto('file://' + FILE); await p.waitForTimeout(3000)

  const res = await p.evaluate((only) => {
    const rooms = Object.keys(window)
      .filter(k => /_ROOM$/.test(k) && window[k] && Array.isArray(window[k].plan))
      .filter(k => !only || k.toLowerCase().includes(only.toLowerCase()))
    const out = []
    for (const key of rooms) {
      const room = window[key]
      const g = window.roomGrid(room, true)
      const { G, COLS, ROWS, CELL } = g
      const free = (c, r) => c >= 0 && r >= 0 && c < COLS && r < ROWS && !G[r * COLS + c]

      // 哪件家具占着这一点 —— 只说「站进家具里」没用,要说清是哪一件
      const A = window.ASSETS || {}
      const whoAt = (x, y) => (room.plan || []).filter(pl => {
        const a = A[pl[0]]
        if (!a || !a.foot || a.walkable) return false
        const [ox, oy, fw, fh] = a.foot
        if (!fw || !fh) return false
        return x >= pl[1] + ox - 8 && x <= pl[1] + ox + fw + 8 &&
               y >= pl[2] + oy - 8 && y <= pl[2] + oy + fh + 8
      }).map(pl => pl[0])

      /* 「他实际会停在离锚点多远的地方」—— 不要自己造可达性判据。
         引擎的 gridPath 在目标不可达时用 near() 退到最近的可达格,所以
         锚点【本来就允许】落在障碍格上:走到桌边坐下、走到床边躺下,锚点
         都在家具的 NAV_PAD(16px)膨胀区里。把那个当穿模是错判。
         自己泛洪也不行 —— 第一版从最下面一排起泛,起点落在房间的延伸带上
         (room.extBand,门外那截),跟房内不连通,于是阿云房 18 个锚点报了 17 个。
         正解:直接问引擎走一趟,看它把人放在哪儿,差多少就是差多少。 */
      const dropDist = (sx, sy, tx, ty) => {
        // 同一格来回:gridPath 返回 null,那不是「走不到」,是「不用走」
        if (Math.abs(sx - tx) < CELL && Math.abs(sy - ty) < CELL) return { d: 0 }
        const path = window.gridPath(room, sx, sy, tx, ty)
        if (!path || !path.length) return { d: 9999 }
        const last = path[path.length - 1]
        // 落点一并带出来 —— 报「够不着」而不说该挪到哪儿,等于把活推回给人
        return { d: Math.max(Math.abs(last[0] - tx), Math.abs(last[1] - ty)), at: last }
      }

      // 站进家具【本体】—— 不含 PAD。这一条才是真的「他站在柜子里」
      const insideBody = (x, y) => (room.plan || []).filter(pl => {
        const a = A[pl[0]]
        if (!a || !a.foot || a.walkable) return false
        const [ox, oy, fw, fh] = a.foot
        if (!fw || !fh) return false
        return x > pl[1] + ox && x < pl[1] + ox + fw && y > pl[2] + oy && y < pl[2] + oy + fh
      }).map(pl => pl[0])

      const groups = room.acts || []
      const bad = [], notes = []
      let anchors = 0, paths = 0

      const check = (label, actorId, anchorIsFoot, list) => {
        const F = anchorIsFoot ? [0, 0] : ((window.ACTORS[actorId] && window.ACTORS[actorId].foot) || [0, 0])
        const pts = []
        for (const a of list) {
          if (typeof a.x !== 'number' || typeof a.y !== 'number') continue
          pts.push([a.x + F[0], a.y + F[1], a.id || '?', a])
        }
        if (!pts.length) return
        const [ox, oy] = pts[0]                      // 出发点:该组第一个锚点
        for (const [fx, fy, id, a] of pts) {
          anchors++
          const body = insideBody(fx, fy)
          const r = fx === ox && fy === oy ? { d: 0 } : dropDist(ox, oy, fx, fy)
          const d = r.d
          if (body.length) {
            bad.push(`${label} 锚点「${id}」脚点(${fx},${fy}) 站进了 ${body.join(' · ')} 【里面】`)
          } else if (d > 80) {
            bad.push(`${label} 锚点「${id}」脚点(${fx},${fy}) 他只能停在 ${d}px 开外 —— 周边: ${whoAt(fx, fy).join(' · ') || '(被围死)'}` +
                     (r.at ? `\n        → 引擎实际把他放在 (${r.at[0]},${r.at[1]})，锚点改成这儿就对上了` : ''))
          } else if (d > 40) {
            notes.push(`${label}「${id}」他会停在 ${d}px 外 —— 走到家具边算正常,再远就该挪锚点了`)
          }
          // 睡觉那类会被 sleepAt 挪到床上,床本身是障碍 —— 那是有意的,单列
          if (a.sleepAt) notes.push(`${label}「${id}」sleepAt(${a.sleepAt}) 落在家具上(睡在床上,正常)`)
        }
        // 两两互走一遍,看折线拐点有没有踩进家具本体
        for (let i = 0; i < pts.length; i++) for (let j = 0; j < pts.length; j++) {
          if (i === j) continue
          const [sx, sy, sid] = pts[i], [tx, ty, tid] = pts[j]
          const path = window.gridPath(room, sx, sy, tx, ty)
          paths++
          if (!path) { bad.push(`${label} ${sid} → ${tid} 找不到路`); continue }
          for (const q of path) {
            const hit = insideBody(q[0], q[1])
            if (hit.length) { bad.push(`${label} ${sid} → ${tid} 路径拐点(${q[0]},${q[1]}) 穿过 ${hit.join(' · ')}`); break }
          }
        }
      }

      for (const grp of groups) check(grp.actor, grp.actor, grp.anchorIsFoot, grp.list || [])

      // 表演锚点:玩家点按钮唯一会去的地方
      const pf = room.perform && room.perform.actor
      if (pf && typeof pf.x === 'number') {
        const main = (groups[0] && groups[0].actor) || null
        const F = (main && window.ACTORS[main] && window.ACTORS[main].foot) || [0, 0]
        const fx = pf.x + F[0], fy = pf.y + F[1]
        anchors++
        const g0 = groups[0] && (groups[0].list || []).find(a => typeof a.x === 'number')
        const body = insideBody(fx, fy)
        const d = g0 ? dropDist(g0.x + F[0], g0.y + F[1], fx, fy).d : 0
        if (body.length) bad.push(`表演锚点 脚点(${fx},${fy}) 站进了 ${body.join(' · ')} 【里面】`)
        else if (d > 80) bad.push(`表演锚点 脚点(${fx},${fy}) 他只能停在 ${d}px 开外 —— 玩家点了按钮却看他停在半路`)
      }
      out.push({ key, groups: groups.length, anchors, paths, bad, notes })
    }
    return out
  }, ONLY)

  let total = 0
  for (const r of res) {
    console.log(`── ${r.key}  行为组 ${r.groups} · 锚点 ${r.anchors} · 走了 ${r.paths} 条路`)
    if (!r.groups) {
      console.log('   ~ 这间房没在 ROOM.acts 上挂行为表 —— 查不了。')
      console.log('     在 loop 的首帧接线处(与 wirePerform 并列)挂:')
      console.log('     window.X_ROOM.acts = [{ actor: \'x\', list: ACTS }]')
      continue
    }
    if (r.bad.length) { total += r.bad.length; r.bad.forEach(m => console.log('   ✗ ' + m)) }
    else console.log('   ✓ 锚点与路径全干净')
    r.notes.forEach(m => console.log('   · ' + m))
  }
  console.log('\n' + (total ? `✗ ${total} 处` : '✓ 0 处'))
  console.log('⚠ 报 0 建立在 S5 的 foot 之上 —— 碰撞体画错了,穿模照样过。先 assetlint + roomaudit D。')
  await b.close()
  if (total) process.exit(1)
})()
